### PR TITLE
Improve longer appointment style

### LINKED ISSUE #129: Calendar: Visualize entries with end or start scheduled beyond current view
Here are some mockups for visualizing entries that start or end on a point in time that is not within the view.

### Daily
Frame 20

### Weekly
Frame 22

### e.g. Daily
Frame 24

## Click to reveal details

Frame 27

## Monthly

Frame 28

### PR DESCRIPTION
**Pull request by raviks789**
fixes #129

### CODE REVIEW

File: library/Notifications/Widget/Calendar/BaseGrid.php [hunk 1]
```diff
@@ -289,16 +340,92 @@ protected function assembleGridOverlay(BaseHtmlElement $overlay): void
...
+    /**
+     * Fetch properties for entryHtml
+     *
+     * @param string $entryColor
+     * @param string|null $direction
+     * @param string|null $opposite
+     *
+     * @return array
+     */
+    protected function getEntryHtmlProperties(
```
Comment (nilmerg) [inline library/Notifications/Widget/Calendar/BaseGrid.php:361]: Please define as much as possible in CSS already. The only variables are the contact's color values. Since you need them now for more than just a background and border color, just set some CSS variables and reference them instead of generating the entire rules here. I suppose once you did this, this method is obsolete.

File: library/Notifications/Widget/Calendar/BaseGrid.php [hunk 2]
```diff
@@ -289,16 +340,92 @@ protected function assembleGridOverlay(BaseHtmlElement $overlay): void
                         'data-col-end' => $gridArea[3]
                     ])
                 );
-                $this->assembleEntry($entryHtml, $entry, $continuation);
+
+                $this->assembleEntry($entryHtml, $entry, $startText, $endText, $continuation);
                 $overlay->addHtml($entryHtml);
 
                 $continuation = true;
```
Comment (nilmerg) [inline library/Notifications/Widget/Calendar/BaseGrid.php:347]: Doesn't this already help in deciding whether a gradient is required or not? The only missing thing here, is that an entry which starts outside the current view isn't handled. This only covers a single direction, forward, and needs a counterpart. I didn't check how many of the calculations above are redundant, but since you're not using nor replacing this flag, I think some of them are.

**Review by nilmerg — CHANGES_REQUESTED** [review]

File: public/css/calendar.less [hunk 3]
```diff
@@ -1,4 +1,5 @@
 /* Layout */
+@entry-color: var(--entry-color, @default-bg);
```
Comment (nilmerg) [inline public/css/calendar.less:2]: What is this for? Just use `var(--entry-bg)` where this is used instead.

File: public/css/calendar.less [hunk 4]
```diff
@@ -338,6 +425,10 @@
     }
   }
 
+  .column-title:not(:first-of-type) {
+    border-left: 1px solid @gray-lighter;
```
Comment (nilmerg) [inline public/css/calendar.less:429]: This doesn't align with the grid. Why is this needed?

**Review by nilmerg — CHANGES_REQUESTED** [review]

File: public/css/calendar.less (hunk 4 above)
Comment (raviks789) [reply public/css/calendar.less:429]: I would have to align this, but it is needed. Because in case of Week, it could be confusing to differentiate the recipients if you have many recipients in an entry.

Comment (raviks789) [reply public/css/calendar.less:429]: As discussed offline this change will be removed from this PR, and an [issue](https://github.com/Icinga/icinga-notifications-web/issues/168) has been opened for this.

File: public/css/calendar.less
Comment (nilmerg) [inline public/css/calendar.less:1]: If the description of an entry is long enough, the end date is not visible.

File: public/css/calendar.less [hunk 5]
```diff
@@ -183,6 +188,10 @@
   .grid,
   .overlay {
     border-left: none;
+
+    .title {
+      width: 100%;
```
Comment (nilmerg) [inline public/css/calendar.less:193]: This lets the description begin below the title, needlessly. (Since the end time isn't visible anyway)

File: public/css/calendar.less [hunk 6]
```diff
@@ -258,6 +283,17 @@
 
     .entry {
       pointer-events: all;
+
+      .starts-at,
+      .ends-at {
+        font-weight: normal;
+        color: @text-color-light;
```
Comment (nilmerg) [inline public/css/calendar.less:290]: This is style!

File: public/css/calendar.less
Comment (nilmerg) [inline public/css/calendar.less:1]: I'll fix the commented issues in this file myself. I'd like to get this done asap.

**Review by nilmerg — CHANGES_REQUESTED** [review]

File: public/css/calendar.less [hunk 7]
```diff
@@ -324,6 +408,10 @@
         opacity: .8;
       }
     }
+
+    .description {
+      mix-blend-mode: screen;
```
Comment (nilmerg) [inline public/css/calendar.less]: I've checked this with Florian again, and it seems to have no effect, not for the description nor for the title.

File: library/Notifications/Widget/Calendar/BaseGrid.php [hunk 8]
```diff
@@ -307,7 +358,47 @@ protected function assembleEntry(BaseHtmlElement $html, Entry $entry, bool $isCo
...
+                HtmlElement::create(
+                    'div',
+                    ['class' => 'starts-at'],
+                    $startText
+                )
+            );
+            $titleAttr = $startText . ' ' . $titleAttr;
+        }
+
+        if ($endText) {
```
[the remark this answers is not in the document]
Comment (raviks789) [reply library/Notifications/Widget/Calendar/BaseGrid.php:388]: Here, we do not need an elseif. Imagine if you are in day mode and you have both the start text and the end text (I need to add a continuation type for this `FROM_PREV_TO_NEXT_GRID`). In that case you need them both in the title attribute.

File: library/Notifications/Widget/Calendar/BaseGrid.php [hunk 9]
```diff
@@ -271,34 +285,71 @@ protected function assembleGridOverlay(BaseHtmlElement $overlay): void
...
                         'data-col-end' => $gridArea[3]
                     ])
                 );
-                $this->assembleEntry($entryHtml, $entry, $continuation);
+
+                if ($fromPrevGrid) {
+                    $continuationType = self::FROM_PREV_GRID;
+                } elseif ($toNextGrid) {
+                    $continuationType = self::TO_NEXT_GRID;
+                } else {
```
[the remark this answers is not in the document]
Comment (raviks789) [reply library/Notifications/Widget/Calendar/BaseGrid.php:327]: You are right. It should be only set to `self::TO_NEXT_GRID`, when the entry continues to the next row.

File: library/Notifications/Widget/Calendar/BaseGrid.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (raviks789) [reply library/Notifications/Widget/Calendar/BaseGrid.php:388]: Constant `ACROSS_GRID` has been added to address the entries in day mode, that starts in previous grid and continues to next grid.

File: library/Notifications/Widget/Calendar/BaseGrid.php [hunk 10]
```diff
@@ -271,33 +291,66 @@ protected function assembleGridOverlay(BaseHtmlElement $overlay): void
...
     }
 
-    protected function assembleEntry(BaseHtmlElement $html, Entry $entry, bool $isContinuation): void
+    /**
+     * Assemble the entry in the grid
+     *
+     * @param BaseHtmlElement $html Container where to add the entry's HTML
+     * @param Entry $entry The entry to assemble
+     * @param ?ContinuationType $continuationType Continuation type of the entry's HTML
+     */
```
Comment (nilmerg) [inline library/Notifications/Widget/Calendar/BaseGrid.php:352]: missing return type in doc

**Review by nilmerg — CHANGES_REQUESTED** [review]
please rebase

**Review by nilmerg — APPROVED** [review]
